File: app/src/main/java/com/voiceassist/lixinyu/voiceassist/settings/adapter/NodeListAdapter.java
package com.voiceassist.lixinyu.voiceassist.settings.adapter;

import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.RecyclerView.ViewHolder;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.voiceassist.lixinyu.voiceassist.R;
import com.voiceassist.lixinyu.voiceassist.entity.dto.Node;

import java.util.List;

/**
 * Created by lilidan on 2018/1/22.
 */

public class NodeListAdapter extends RecyclerView.Adapter<NodeListAdapter.MyViewHolder> implements View.OnClickListener {

    private Context mContext;
    private List<Node> mData;

    private OnItemClickListener mOnItemClickListener;

    public NodeListAdapter(Context context, List<Node> data) {
        this.mContext = context;
        this.mData = data;
    }

    @Override
    public MyViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        MyViewHolder holder = new MyViewHolder(LayoutInflater.from(mContext).inflate(R.layout.edit_node_item, parent, false));

        holder.root.setOnClickListener(this);

        return holder;
    }

    @Override
    public void onBindViewHolder(MyViewHolder holder, int position) {
        Node node = mData.get(position);
        if (null != node) holder.tv.setText(node.cnName);

        holder.root.setTag(position);
    }

    @Override
    public int getItemCount() {
        return mData.size();
    }

    @Override
    public void onClick(View v) {
        int position = (int) v.getTag();
        if (null != mOnItemClickListener) {
            mOnItemClickListener.onClick(this, position, mData.get(position));
        }
    }

    class MyViewHolder extends ViewHolder {

        TextView tv;
        LinearLayout root;

        public MyViewHolder(View view) {
            super(view);
            tv = (TextView) view.findViewById(R.id.edit_node_item_textview);
            root = view.findViewById(R.id.edit_node_item_root);
        }
    }

    public interface OnItemClickListener {
        void onClick(NodeListAdapter adapter, int position, Node node);
    }

    public void setOnItemClickListener(OnItemClickListener listener) {
        this.mOnItemClickListener = listener;
    }
}

File: app/src/main/java/com/voiceassist/lixinyu/voiceassist/settings/adapter/NodeListAdapter.kt
package com.voiceassist.lixinyu.voiceassist.settings.adapter

import android.content.Context
import android.support.v7.widget.RecyclerView
import android.support.v7.widget.RecyclerView.ViewHolder
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.LinearLayout
import android.widget.TextView

import com.voiceassist.lixinyu.voiceassist.R
import com.voiceassist.lixinyu.voiceassist.entity.dto.Node

/**
 * Created by lilidan on 2018/1/22.
 */

internal class NodeListAdapter(protected var mContext: Context, protected var mData: List<Node>?) : RecyclerView.Adapter<NodeListAdapter.MyViewHolder>(), View.OnClickListener, View.OnLongClickListener {

    private var mOnItemClickListener: OnItemClickListener? = null

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): MyViewHolder {
        val holder = MyViewHolder(LayoutInflater.from(mContext).inflate(R.layout.item_edit_node, parent, false))

        holder.itemView.setOnClickListener(this)
        holder.itemView.setOnLongClickListener(this)

        return holder
    }

    override fun onBindViewHolder(holder: MyViewHolder, position: Int) {
        val node = mData?.get(position)
        if (null != node) holder.tv.text = node.cnName

        holder.itemView.tag = position
    }

    override fun getItemCount(): Int {
        return mData?.size ?: 0
    }

    override fun onClick(v: View) {
        val position = v.tag as Int
        if (null != mOnItemClickListener) {
            mOnItemClickListener!!.onClick(this, position, mData?.get(position))
        }
    }

    override fun onLongClick(v: View): Boolean {
        val position = v.tag as Int

        if (null != mOnItemClickListener) {
            mOnItemClickListener!!.onLongClick(this, position, mData?.get(position))
        }

        return false
    }

    internal inner class MyViewHolder(view: View) : ViewHolder(view) {

        var tv: TextView
        var root: LinearLayout

        init {
            tv = view.findViewById<View>(R.id.edit_node_item_textview) as TextView
            root = view.findViewById(R.id.edit_node_item_root)
        }
    }

    interface OnItemClickListener {
        fun onClick(adapter: NodeListAdapter, position: Int, node: Node?)
        fun onLongClick(adapter: NodeListAdapter, position: Int, node: Node?)
    }

    fun setOnItemClickListener(listener: OnItemClickListener) {
        this.mOnItemClickListener = listener
    }
}
